profiling/src/collections/identifiable/string_table.rs

```diff
@@ -0,0 +1,266 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2023-Present Datadog, Inc.
+
+use super::*;
+use bumpalo::Bump;
+use ouroboros::self_referencing;
+use std::collections::HashMap;
+use std::hash::BuildHasherDefault;
+
+#[cfg(test)]
+use std::ops::Range;
+
+struct BorrowedStringTable<'b> {
+    /// The arena to store the characters in.
+    arena: &'b Bump,
+
+    /// Used to have efficient lookup by [StringId], and to provide an
+    /// [Iterator] over the strings.
+    vec: Vec<&'b str>,
+
+    /// Used to have efficient lookup by [&str].
+    map: HashMap<&'b str, StringId, BuildHasherDefault<rustc_hash::FxHasher>>,
+}
+
+#[self_referencing]
+struct StringTableCell {
+    // This arena holds the characters of the strings. The memory used to hold
+    // the vec and map to implement the lookups required are not in the arena,
+    // because they would be re-allocated over time but the previous data
+    // wouldn't generally get reclaimed. This makes them a poor fit for an
+    // arena allocator.
+    owner: Bump,
+
+    // This says that the BorrowedStringTable will hold a reference to the
+    // field `owner`. Rust does not allow this as there are many ways to make
+    // this code unsafe. The ouroboros crate is used to provide a safe
+    // abstraction for a subset of self-referential behavior.
+    #[borrows(owner)]
+    #[covariant]
+    dependent: BorrowedStringTable<'this>,
+}
+
+/// The [StringTable] stores strings and associates them with [StringId]s,
+/// which correspond to the order in which strings were inserted. The empty
+/// string is always associated with [StringId::ZERO].
+pub struct StringTable {
+    // ouroboros will add a lot of functions to this struct, which we don't
+    // want to expose publicly, so the internals are wrapped and private.
+    inner: StringTableCell,
+}
+
+impl StringTable {
+    // Not guaranteed for a given system, but a very common size.
+    const PAGE_SIZE: usize = 4096;
```

Comment:
We could but it doesn't really matter because we're too far away in abstraction levels. Just really just documenting some of the choices of magic numbers.